core/vm/wasm_qei.go

```diff
@@ -0,0 +1,894 @@
+	"getGasLeft",
+	"getBlockGasLimit",
+	"getTxGasPrice",
+	"log",
+	"getBlockNumber",
+	"getTxOrigin",
+	"finish",
+	"revert",
+	"getReturnDataSize",
+	"returnDataCopy",
+	"selfDestruct",
+	"getBlockTimestamp",
+}
+
+func InstantiateWASMRuntime(ctx context.Context, in *WASMInterpreter) wazero.Runtime {
+
+	// Create a new WebAssembly Runtime.
+	r := wazero.NewRuntime(ctx)
+
+	_, errEnv := r.NewHostModuleBuilder("env").
+		NewFunctionBuilder().WithFunc(logUint32).Export("hostLogUint32").
+		NewFunctionBuilder().WithFunc(logString).Export("hostLogString").
+		NewFunctionBuilder().WithFunc(in.useGas).Export("useGas").
+		NewFunctionBuilder().WithFunc(in.getAddress).Export("getAddress").
+		NewFunctionBuilder().WithFunc(in.getExternalBalance).Export("getExternalBalance").
+		NewFunctionBuilder().WithFunc(in.getCaller).Export("getCaller").
+		NewFunctionBuilder().WithFunc(in.getBlockHash).Export("getBlockHash").
+		NewFunctionBuilder().WithFunc(in.getBlockNumber).Export("getBlockNumber").
+		NewFunctionBuilder().WithFunc(in.getBlockTimestamp).Export("getBlockTimestamp").
+		Instantiate(ctx)
+	if errEnv != nil {
+		fmt.Println("🔴 Error with env:", errEnv)
+	}
+
+	_, errInstantiate := wasi_snapshot_preview1.Instantiate(ctx, r)
+	if errInstantiate != nil {
+		fmt.Println("🔴 Error with Instantiate:", errInstantiate)
+	}
+
+	fmt.Println("🤖: Instantiated WASM Runtime")
+
+	return r
+}
+
+func logUint32(value uint32) {
+	fmt.Println("🤖:", value)
+}
+
+func logString(ctx context.Context, module api.Module, offset, byteCount uint32) {
+	buf, ok := module.Memory().Read(offset, byteCount)
+	if !ok {
+		log.Panicf("🟥 Memory.Read(%d, %d) out of range", offset, byteCount)
+	}
+	fmt.Println("👽:", string(buf))
+}
+
+func (in WASMInterpreter) gasAccounting(cost uint64) {
+	if in.contract == nil {
+		panic("nil contract")
+	}
+	if cost > in.contract.Gas {
+		panic(fmt.Sprintf("out of gas %d > %d", cost, in.contract.Gas))
+	}
+	in.contract.Gas -= cost
+}
+
+func (in *WASMInterpreter) useGas(ctx context.Context, module api.Module, amount int64) {
+	fmt.Println("🤖: useGas", amount)
+	fmt.Println("🤖: useGas", in.contract.Gas)
+	in.gasAccounting(uint64(amount))
+}
+
+func readSize(ctx context.Context, module api.Module, offset uint32, size uint32) []byte {
+	// TODO modify the process interface to find out how much memory is
+	// available on the system.
+	bytes, ok := module.Memory().Read(offset, size)
+	if !ok {
+		log.Panicf("🟥 Memory.ReadInto(%d, %d) out of range", offset, size)
+	}
+
+	return bytes
+}
+
+func writeBytes(ctx context.Context, module api.Module, bytes []byte, offset uint32) {
+	if !module.Memory().Write(offset, bytes) {
+		log.Panicf("🟥 Memory.Write(%d, %d) out of range", offset, len(bytes))
+	}
+}
+
+func swapEndian(src []byte) []byte {
+	ret := make([]byte, len(src))
+	for i, v := range src {
+		ret[len(src)-i-1] = v
+	}
+	return ret
+}
+
+func (in *WASMInterpreter) getAddress(ctx context.Context, module api.Module, resultOffset uint32) {
```

Comment:
Would be nice if these methods were part of an interface defined in WASMInterpreter

Reply:
Lfg

Reply:
when will the mainnet network launch